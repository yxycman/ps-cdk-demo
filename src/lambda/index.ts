import { Stack, StackProps, Stage, StageProps } from 'aws-cdk-lib';
import { Code, Runtime, Function as Function } from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';
import { CodePipelineSource, CodePipeline, ShellStep } from 'aws-cdk-lib/pipelines';


export interface ILambdaHelloWorldPipelineStackProps extends StackProps {
  /**
   * GitHub Repository string.
   */
  gitHubRepositoryString: string;

  /**
   * LambdaHelloWorld deployment configuration.
   */
  lambdaHelloWorldDeploymentConfig: ILambdaHelloWorldStackProps[];
}


export interface ILambdaHelloWorldPipelineStageProps extends StageProps {
  /**
   * LambdaHelloWorld deployment configuration.
   */
  lambdaHelloWorldDeploymentConfig: ILambdaHelloWorldStackProps[];
}


export interface ILambdaHelloWorldStackProps extends StackProps {
  /**
   * LambdaHelloWorld message.
   */
  lambdaHelloWorldMessage: string;
}


export class LambdaHelloWorldStack extends Stack {
  constructor(scope: Construct, id: string, props: ILambdaHelloWorldStackProps) {
    super(scope, id, props);

    new Function(this, 'HelloWorldFunction', {
      runtime: Runtime.NODEJS_20_X,
      code: Code.fromAsset('./src/lambda/assets'),
      handler: 'hw.handler',
      environment: { MESSAGE: props.lambdaHelloWorldMessage},
    });
  }
}


// Stage class to hold the LambdaHelloWorld stack
export class LambdaHelloWorldPipelineStage extends Stage {
  constructor(scope: Construct, id: string, props: ILambdaHelloWorldPipelineStageProps) {
    super(scope, id, props);

    props.lambdaHelloWorldDeploymentConfig.forEach((deploymentConfig) => {
      new LambdaHelloWorldStack(this, `LambdaHelloWorld-${deploymentConfig.env?.account}-${deploymentConfig.env?.region}`, {
        stackName: `LambdaHelloWorld-${deploymentConfig.env?.account}-${deploymentConfig.env?.region}`,
        description: 'The LambdaHelloWorld stack',
        env: deploymentConfig.env,
        lambdaHelloWorldMessage: deploymentConfig.lambdaHelloWorldMessage,
      });
    });
  }
}


// Stack class to hold the pipeline for LambdaHelloWorld
export class LambdaHelloWorldPipelineStack extends Stack {
  constructor(scope: Construct, id: string, props: ILambdaHelloWorldPipelineStackProps) {
    super(scope, id, props);

    const pipeline = new CodePipeline(this, 'LambdaHelloWorldPipeline', {
      crossAccountKeys: true,
      synth: new ShellStep('Synth', {
        input: CodePipelineSource.gitHub(props.gitHubRepositoryString, 'main'),
        // Installing and building stack with projen
        commands: [
          'n auto',
          'npm ci',
          'npx projen build',
          'npx projen synth',
        ],
      }),

      // codeBuildDefaults: {
      //   partialBuildSpec: BuildSpec.fromObject({
      //     version: '0.2',
      //     phases: {
      //       install: {
      //         'runtime-versions': {
      //           nodejs: '18',
      //         },
      //       },
      //     },
      //   }),

      //   buildEnvironment: {
      //     buildImage: LinuxBuildImage.STANDARD_7_0,

      //     environmentVariables: {
      //       NODE_CONFIG_ENV: {
      //         type: BuildEnvironmentVariableType.PLAINTEXT,
      //         value: "ENV",
      //       },
      //       AWS_ACCOUNT_ID: {
      //         type: BuildEnvironmentVariableType.PLAINTEXT,
      //         value: props.env?.account,
      //       },
      //       NODE_AUTH_TOKEN: {
      //         type: BuildEnvironmentVariableType.SECRETS_MANAGER,
      //         value: 'NpmToken',
      //       },
      //     },
      //   },
      // },
    });

    const LambdaHelloWorldWave = pipeline.addWave('LambdaHelloWorld');
    LambdaHelloWorldWave.addStage(
      new LambdaHelloWorldPipelineStage(this, 'LambdaHelloWorld', {
        env: props.env,
        lambdaHelloWorldDeploymentConfig: props.lambdaHelloWorldDeploymentConfig,
      })
    );

    pipeline.buildPipeline();
  }
}
