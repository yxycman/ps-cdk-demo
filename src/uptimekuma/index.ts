export * from './pipeline';
export * from './stack';