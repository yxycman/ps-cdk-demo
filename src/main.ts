import { App } from 'aws-cdk-lib';
import { LambdaHelloWorldPipelineStack, LambdaHelloWorldStack } from './lambda';
//import { UptimeKumaStack } from './uptimekuma';

const app = new App();
// for development, use account/region from cdk cli
const devEnv = { account: process.env.CDK_DEFAULT_ACCOUNT, region: process.env.CDK_DEFAULT_REGION };

new LambdaHelloWorldStack(app, 'LambdaHelloWorld', {
  description: 'Demo CDK Stack',
  env: devEnv,
  lambdaHelloWorldMessage: 'ProfiSea'
});

new LambdaHelloWorldPipelineStack(app, 'LambdaHelloWorldPipeline', {
  description: 'Demo CDK Stack',
  gitHubRepositoryString: 'yxycman/ps-cdk-demo',
  env: devEnv,
  lambdaHelloWorldDeploymentConfig: [
    {
      env: { account: process.env.CDK_DEFAULT_ACCOUNT, region: 'us-east-1' },
      lambdaHelloWorldMessage: 'ProfiSea in us-east-1',
    },
    {
      env: { account: process.env.CDK_DEFAULT_ACCOUNT, region: 'us-east-2' },
      lambdaHelloWorldMessage: 'ProfiSea in us-east-2',
    },
  ],
});

// new UptimeKumaStack(app, 'UptimeKumaStack', {
//   description: 'UptimeKuma CDK Stack',
//   env: devEnv,
//   instanceType: 't3.large',
//   vpcId: 'vpc-0c2485d43df151fed',
// });

app.synth();