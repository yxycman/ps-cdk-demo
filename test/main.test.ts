import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { LambdaHelloWorldStack } from '../src/lambda';

test('Snapshot', () => {
  const app = new App();
  const stack = new LambdaHelloWorldStack(app, 'test', {
    description: 'LambdaHelloWorld CDK Stack',
    env: { account: '123456789012', region: 'us-east-1' },
    lambdaHelloWorldMessage: 'test',
  });

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});