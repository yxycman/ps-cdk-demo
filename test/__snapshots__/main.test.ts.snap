// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Description": "UptimeKuma CDK Stack",
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "SsmParameterValueawsserviceecsoptimizedamiamazonlinux2recommendedimageidC96584B6F00A464EAD1953AFF4B05118Parameter": {
      "Default": "/aws/service/ecs/optimized-ami/amazon-linux-2/recommended/image_id",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::Image::Id>",
    },
  },
  "Resources": {
    "ASG46ED3070": {
      "Properties": {
        "LaunchTemplate": {
          "LaunchTemplateId": {
            "Ref": "LaunchTemplate04EC5460",
          },
          "Version": {
            "Fn::GetAtt": [
              "LaunchTemplate04EC5460",
              "LatestVersionNumber",
            ],
          },
        },
        "MaxSize": "1",
        "MinSize": "1",
        "NewInstancesProtectedFromScaleIn": true,
        "VPCZoneIdentifier": [
          "p-12345",
        ],
      },
      "Type": "AWS::AutoScaling::AutoScalingGroup",
      "UpdatePolicy": {
        "AutoScalingScheduledAction": {
          "IgnoreUnmodifiedGroupSizeProperties": true,
        },
      },
    },
    "AlbSecurityGroup86A59E99": {
      "Properties": {
        "GroupDescription": "test/AlbSecurityGroup",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": "vpc-12345",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "AsgCapacityProvider760D11D9": {
      "Properties": {
        "AutoScalingGroupProvider": {
          "AutoScalingGroupArn": {
            "Ref": "ASG46ED3070",
          },
          "ManagedScaling": {
            "Status": "ENABLED",
            "TargetCapacity": 100,
          },
          "ManagedTerminationProtection": "ENABLED",
        },
      },
      "Type": "AWS::ECS::CapacityProvider",
    },
    "AsgSecurityGroupABD137B7": {
      "Properties": {
        "GroupDescription": "test/AsgSecurityGroup",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": [
          {
            "Description": "ALB Access",
            "FromPort": 0,
            "IpProtocol": "tcp",
            "SourceSecurityGroupId": {
              "Fn::GetAtt": [
                "AlbSecurityGroup86A59E99",
                "GroupId",
              ],
            },
            "ToPort": 65535,
          },
        ],
        "VpcId": "vpc-12345",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "Cluster3DA9CCBA": {
      "Properties": {
        "CapacityProviders": [
          {
            "Ref": "AsgCapacityProvider760D11D9",
          },
        ],
        "Cluster": {
          "Ref": "ClusterEB0386A7",
        },
        "DefaultCapacityProviderStrategy": [],
      },
      "Type": "AWS::ECS::ClusterCapacityProviderAssociations",
    },
    "ClusterEB0386A7": {
      "Type": "AWS::ECS::Cluster",
    },
    "EC2RoleDefaultPolicy64E44857": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "ecs:DeregisterContainerInstance",
                "ecs:RegisterContainerInstance",
                "ecs:Submit*",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "ClusterEB0386A7",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "ecs:Poll",
                "ecs:StartTelemetrySession",
              ],
              "Condition": {
                "ArnEquals": {
                  "ecs:cluster": {
                    "Fn::GetAtt": [
                      "ClusterEB0386A7",
                      "Arn",
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "ecs:DiscoverPollEndpoint",
                "ecr:GetAuthorizationToken",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "EC2RoleDefaultPolicy64E44857",
        "Roles": [
          {
            "Ref": "EC2RoleF978FC1C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "EC2RoleF978FC1C": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonSSMManagedEC2InstanceDefaultPolicy",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonEC2ContainerServiceforEC2Role",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ECSTaskRoleF2ADB362": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "LaunchTemplate04EC5460": {
      "DependsOn": [
        "EC2RoleDefaultPolicy64E44857",
        "EC2RoleF978FC1C",
      ],
      "Properties": {
        "LaunchTemplateData": {
          "IamInstanceProfile": {
            "Arn": {
              "Fn::GetAtt": [
                "LaunchTemplateProfile94AA77CE",
                "Arn",
              ],
            },
          },
          "ImageId": {
            "Ref": "SsmParameterValueawsserviceecsoptimizedamiamazonlinux2recommendedimageidC96584B6F00A464EAD1953AFF4B05118Parameter",
          },
          "InstanceType": "t3.medium",
          "SecurityGroupIds": [
            {
              "Fn::GetAtt": [
                "AsgSecurityGroupABD137B7",
                "GroupId",
              ],
            },
          ],
          "TagSpecifications": [
            {
              "ResourceType": "instance",
              "Tags": [
                {
                  "Key": "Name",
                  "Value": "test/LaunchTemplate",
                },
              ],
            },
            {
              "ResourceType": "volume",
              "Tags": [
                {
                  "Key": "Name",
                  "Value": "test/LaunchTemplate",
                },
              ],
            },
          ],
          "UserData": {
            "Fn::Base64": {
              "Fn::Join": [
                "",
                [
                  "#!/bin/bash
echo ECS_CLUSTER=",
                  {
                    "Ref": "ClusterEB0386A7",
                  },
                  " >> /etc/ecs/ecs.config
sudo iptables --insert FORWARD 1 --in-interface docker+ --destination 169.254.169.254/32 --jump DROP
sudo service iptables save
echo ECS_AWSVPC_BLOCK_IMDS=true >> /etc/ecs/ecs.config",
                ],
              ],
            },
          },
        },
        "LaunchTemplateName": "UptimeKumaLaunchTemplate",
        "TagSpecifications": [
          {
            "ResourceType": "launch-template",
            "Tags": [
              {
                "Key": "Name",
                "Value": "test/LaunchTemplate",
              },
            ],
          },
        ],
      },
      "Type": "AWS::EC2::LaunchTemplate",
    },
    "LaunchTemplateProfile94AA77CE": {
      "Properties": {
        "Roles": [
          {
            "Ref": "EC2RoleF978FC1C",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "LogGroupF5B46931": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "RetentionInDays": 30,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "LogGroupPolicyCFDD4E41": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutLogEvents",
                "logs:CreateLogStream",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "LogGroupF5B46931",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogGroupPolicyCFDD4E41",
        "Roles": [
          {
            "Ref": "ECSTaskRoleF2ADB362",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "Task79114B6B": {
      "Properties": {
        "ContainerDefinitions": [
          {
            "Essential": true,
            "Image": "louislam/uptime-kuma",
            "LogConfiguration": {
              "LogDriver": "awslogs",
              "Options": {
                "awslogs-group": {
                  "Ref": "LogGroupF5B46931",
                },
                "awslogs-region": "us-east-1",
                "awslogs-stream-prefix": "main",
              },
            },
            "Memory": 3072,
            "Name": "Container",
            "PortMappings": [
              {
                "ContainerPort": 3001,
                "HostPort": 0,
                "Protocol": "tcp",
              },
            ],
          },
        ],
        "ExecutionRoleArn": {
          "Fn::GetAtt": [
            "TaskExecutionRoleEF964934",
            "Arn",
          ],
        },
        "Family": "testTask5E55995A",
        "NetworkMode": "bridge",
        "RequiresCompatibilities": [
          "EC2",
        ],
        "TaskRoleArn": {
          "Fn::GetAtt": [
            "ECSTaskRoleF2ADB362",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ECS::TaskDefinition",
    },
    "TaskExecutionRoleDefaultPolicy72339918": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "LogGroupF5B46931",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TaskExecutionRoleDefaultPolicy72339918",
        "Roles": [
          {
            "Ref": "TaskExecutionRoleEF964934",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TaskExecutionRoleEF964934": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
